pkg/rules/signature/signature.go

```diff
@@ -61,97 +63,109 @@ func Find(target string, partialEval bool, rulesDir string, rules []string, aioE
+
+					p, err := plugin.Open(path)
+					if err != nil {
+						logger.Error("opening plugin " + path + ": " + err.Error())
+						return err
+					}
+					export, err := p.Lookup("ExportedSignatures")
+					if err != nil {
+						logger.Error("missing Export symbol in plugin " + d.Name())
+						return err
+					}
+					sigs := *export.(*[]detect.Signature)
+					res = append(res, sigs...)
+					return nil
+				})
 			return err
-		}
-
-		if d.IsDir() || filepath.Ext(d.Name()) != ".so" {
-			return nil
-		}
+		},
+		cap.DAC_OVERRIDE,
+	)
 
-		p, err := plugin.Open(path)
-		if err != nil {
-			logger.Error("opening plugin " + path + ": " + err.Error())
-			return err
-		}
-		export, err := p.Lookup("ExportedSignatures")
-		if err != nil {
-			logger.Error("missing Export symbol in plugin " + d.Name())
-			return err
-		}
-		sigs := *export.(*[]detect.Signature)
-		res = append(res, sigs...)
-		return nil
-	})
 	return res, nil
 }
 
 func findRegoSigs(target string, partialEval bool, dir string, aioEnabled bool) ([]detect.Signature, error) {
+	var res []detect.Signature
+
 	modules := make(map[string]string)
 	modules["helper.rego"] = embedded.RegoHelpersCode
 
 	regoHelpers := []string{embedded.RegoHelpersCode}
-	filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
-
-		if d.IsDir() || d.Name() == "helpers.rego" {
-			return nil
-		}
-
-		if !isHelper(d.Name()) {
-			return nil
-		}
-
-		helperCode, err := ioutil.ReadFile(path)
-		if err != nil {
-			logger.Error("reading file " + path + ": " + err.Error())
-			return nil
-		}
-
-		regoHelpers = append(regoHelpers, string(helperCode))
-		modules[path] = string(helperCode)
-		return nil
-	})
-
-	var res []detect.Signature
-	filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
-
-		if d.IsDir() || !isRegoFile(d.Name()) || isHelper(d.Name()) {
-			return nil
-		}
 
-		regoCode, err := ioutil.ReadFile(path)
-		if err != nil {
-			logger.Error("reading file " + path + ": " + err.Error())
-			return nil
-		}
-		modules[path] = string(regoCode)
-		if aioEnabled {
-			return nil
-		}
-		sig, err := regosig.NewRegoSignature(target, partialEval, append(regoHelpers, string(regoCode))...)
-		if err != nil {
-			newlineOffset := bytes.Index(regoCode, []byte("\n"))
-			if newlineOffset == -1 {
-				codeLength := len(regoCode)
-				if codeLength < 22 {
-					newlineOffset = codeLength
-				} else {
-					newlineOffset = 22
+	capabilities.GetInstance().Requested(
+		func() error {
+			filepath.WalkDir(dir,
+				func(path string, d fs.DirEntry, err error) error {
+					if err != nil {
+						return err
+					}
+					if d.IsDir() || d.Name() == "helpers.rego" {
+						return nil
+					}
+					if !isHelper(d.Name()) {
+						return nil
+					}
+					helperCode, err := ioutil.ReadFile(path)
+					if err != nil {
+						logger.Error("reading file " + path + ": " + err.Error())
+						return nil
```

Comment:
What about returning the error in those WalkDir calls? It already does return in case of err not nil and different of SkipDir.

https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/path/filepath/path.go;l=486

Reply:
I did not change that code, I have just wrapped it into a raised privilege callback. I would guess the author didn't want to fully fail if not able to read one rego signature... others would be attempted.

Still, the focus in the PR should be about raising the privilege of already existing code I suppose. We should do other PR for changing existing code IMO.

Reply:
out of scope.